test/functorch/test_control_flow_cuda_initialization.py

```diff
@@ -0,0 +1,78 @@
+# Owner(s): ["module: functorch"]
+import unittest
+
+import torch
+from torch.testing._internal.common_utils import (
+    run_tests,
+    TEST_CUDA_GRAPH_CONDITIONAL_NODES,
+    TestCase,
+)
+
+
+@unittest.skipIf(
+    not TEST_CUDA_GRAPH_CONDITIONAL_NODES,
+    "CUDA 12.4 or greater is required for CUDA Graphs with conditional nodes",
+)
+class TestControlFlowInCUDAGraphInitialization(TestCase):
+    # Duplicated from test_cuda_primary_ctx.py
+    CTX_ALREADY_CREATED_ERR_MSG = (
+        "Tests defined in TestControlFlowInCUDAGraphInitialization must be run in a process "
+        "where CUDA contexts are never created. Use either run_test.py or add "
+        "--subprocess to run each test in a different subprocess."
+    )
+
+    def setUp(self):
+        # Ensure context has not been created beforehand
+        self.assertFalse(
+            torch._C._cuda_hasPrimaryContext(0),
```

Comment:
This is failing when I call `python test/run_test.py -i functorch/test_control_flow_cuda_initialization`. There shouldn't be a context before any cuda operations are run. Not quite sure what's going on yet.

Reply:
CUDA context is being initialized right now because of a torch.cuda.is_available() all that I added here:

```
  File "/home/dgalvez/scratch/code/asr/pytorch/torch/cuda/__init__.py", line 129, in is_available
    return torch._C._cuda_getDeviceCount() > 0
  File "/home/dgalvez/scratch/code/asr/pytorch/torch/_dynamo/variables/torch_function.py", line 153, in __torch_function__
    if torch.cuda.is_available():
  <built-in method abs of module object at remote 0x7ffff45da110>
  File "/home/dgalvez/scratch/code/asr/pytorch/torch/_dynamo/variables/torch_function.py", line 173, in <lambda>
    (lambda o: o(inp0), un_ops),
  File "/home/dgalvez/scratch/code/asr/pytorch/torch/_dynamo/variables/torch_function.py", line 181, in populate_builtin_to_tensor_fn_map
    setup_fn(op)
  File "/home/dgalvez/scratch/code/asr/pytorch/torch/_dynamo/variables/torch_function.py", line 206, in <module>
    populate_builtin_to_tensor_fn_map()
```

It's calling cuInit() during torch import time unfortunately. That code was already a bit sketchy, so I will see if I can get rid of it (it seems that sometimes that code would run on cuda tensors based on my previous experience, but I don't see how that is possible and unfortunately I didn't make a test case to reproduce it.)

Reply:
I fixed the problem in this commit: https://github.com/pytorch/pytorch/pull/140979/commits/1bc9802c8122349034b6c74da267f3de367cefd1 

It is easier to see the change here: https://github.com/pytorch/pytorch/pull/140979/files#diff-b9390381059781e43ca874273a63ac3525d990c9c4d6bb048ef621bbf2a768a2, so you don't have to look at my original changes, which I have reverted.

@mlazos can you let me know whether this works for you? Is there any reason why you used the default torch device rather than cpu device? If the default device is cuda, then there can be trouble if cuda graph capture is running when `torch/_dynamo/variables/torch_function.py` is first imported.

Copy-pasting my commit message:

Make sure that populate_builtin_to_tensor_fn_map() uses CPU ops.                                                                                                                                            

The following program used to fail:
   
```
import torch

torch.set_default_device("cuda")

def f(pred, x):
    return torch.cond(pred, lambda y: torch.sum(y), lambda y: torch.sum(torch.stft(y, 512, return_complex=False)), [x])

g = torch.cuda.CUDAGraph()

pred = torch.tensor(True)
x = torch.ones(1024 * 1024)

with torch.cuda.graph(g, capture_error_mode="thread_local"):
    f(pred, x)
```

torch.set_default_device("cuda") would make the kernels used by
populate_builtin_to_tensor_fn_map() run on the cuda stream used for
stream capture, which would cause a crash for operator.not_, which
call cudaStreamSynchronize(). Even if this weren't the case, all of  
these operations would be unnecessarily captured into the cuda graph,
which reduces performance when replaying the cuda graph.

The dispatch key does not influence the function name, so I just force
dispatching on the CPU to fix this.

Reply:
I think this is fine, just ensure the `__torch_function__` mode tests still pass. 
pytorch/test/dynamo/test_modes.py